### ISSUE TITLE
meta: Support `@protected` for classes

### ISSUE THREAD

**Issue by f-person** [dart-lang/sdk#47952]
It will be nice to be able to annotate classes with `@protected` and gain the benefits we have for class fields.

### Context
I needed to get the state of a `StatefulWidget` (to be more precise, some of its fields) in tests. For that purpose, I had to make the state class public. However the state is intended to only be used in the `StatefulWidget`. Being able to annotate the class with `@protected` and getting analyzer warnings when trying to use it outside of the library it belongs to or tests will help to solve this issue and prevent others from doing it.

**Comment by bwilkerson** [comment]
The purpose of the `protected` annotation is to mark instance members as being intended for use by subclasses but nowhere else. That doesn't really apply to classes, because they're not instance members.

But the `visibleForTesting` annotation appears to have the semantics that you're looking for. Would that annotation work for you?

**Comment by f-person** [comment]
The analyzer gives no warnings when instantiating a class that's marked with `visibleForTesting` outside of its library

**Comment by f-person** [comment]
The only thing I want is making the analyzer show warnings when instantiating a class outside of its library. The annotation used is not important for me :)

**Comment by bwilkerson** [comment]
I think that invoking the constructor ought to constitute a reference to the class for the purposes of `visibleForTesting`, so I think the bug here is the lack of such a diagnostic.

**Comment by bwilkerson** [comment]
@srawlins

**Comment by f-person** [comment]
BTW annotating the constructor itself with `visibleForTesting` results in expected behavior

**Comment by srawlins** [comment]
I've changed the title accordingly. Good catch, @f-person .

**Comment by asashour** [comment]
https://dart-review.googlesource.com/c/sdk/+/242286

**Comment by bwilkerson** [comment]
Thanks for working on this! Unfortunately, that change turns out to be a breaking change in existing code. I don't know yet whether the existing violations are intentional or accidental, but they'll need to be addressed before we can land that change. I'm guessing that people have marked classes as `visibleForTesting` and then inadvertently referenced members because the analyzer didn't catch the error.

**Comment by haodong404** [comment]
Any updates?

**Comment by srawlins** [comment]
No updates.

**Comment by srawlins** [comment]
Duplicate of https://github.com/dart-lang/sdk/issues/50210; that one is newer, but has more discussion.
